tfjs-backend-webgpu/src/binary_op_webgpu.ts

```diff
@@ -15,48 +15,132 @@
  */
 
-import {backend_util} from '@tensorflow/tfjs-core';
+import {backend_util, util} from '@tensorflow/tfjs-core';
+
 import {BinaryOpType, getBinaryOpString} from './binary_op_util';
 import {getMainHeaderAndGlobalIndexString, WebGPUProgram} from './webgpu_program';
 import {computeDispatch, flatDispatchLayout} from './webgpu_util';
 
 export class BinaryOpProgram implements WebGPUProgram {
+  dispatch: [number, number, number];
+  dispatchLayout: {x: number[]};
+  isVec4: boolean;
+  op: BinaryOpType;
   outputShape: number[];
   shaderKey: string;
-  dispatchLayout: {x: number[]};
-  dispatch: [number, number, number];
+  size = true;
+  userCode: string;
   variableNames = ['A', 'B'];
   workGroupSize: [number, number, number];
-  op: BinaryOpType;
-  size = true;
+  workPerThread: number;
 
   constructor(op: BinaryOpType, aShape: number[], bShape: number[]) {
-    // TODO(jiajia.qin@intel.com): Heuristically select a good work group size.
-    const workGroupSizeX = 128;
-    this.workGroupSize = [workGroupSizeX, 1, 1];
     this.outputShape = backend_util.assertAndGetBroadcastShape(aShape, bShape);
     this.dispatchLayout = flatDispatchLayout(this.outputShape);
-
-    this.dispatch = computeDispatch(
-        this.dispatchLayout, this.outputShape, this.workGroupSize);
-    this.shaderKey = `binary_${op}`;
     this.op = op;
-  }
 
-  getUserCode(): string {
-    const opStr = getBinaryOpString(this.op, false);
-    const userCode = `
-      fn binaryOperation(a : f32, b : f32) -> f32 {
-        ${opStr}
+    const useSharedMemoryWithA =
+        aShape.length === 1 && bShape.length > 1 && aShape[0] < 1024;
+    const useSharedMemoryWithB =
+        bShape.length === 1 && aShape.length > 1 && bShape[0] < 1024;
+
+    if (useSharedMemoryWithA || useSharedMemoryWithB) {
+      this.isVec4 = false;
+      // This is an experimental value when using shared memory.
+      // Note that the maximum of workgroup X dimension is 256.
+      this.workGroupSize = [256, 1, 1];
+
+      // lastDimensionSize is used as sharedBuf array size, so can not be
+      // used as uniform.
+      const lastDimensionSize = useSharedMemoryWithB ? bShape[0] : aShape[0];
+      if (lastDimensionSize < 256) {
+        this.workPerThread = 1;
+      } else if (lastDimensionSize < 512) {
+        this.workPerThread = 2;
+      } else {
+        this.workPerThread = 4;
       }
-      ${getMainHeaderAndGlobalIndexString()}
-        if (index < uniforms.size) {
-          let a = getAByOutputIndex(index);
-          let b = getBByOutputIndex(index);
-          setOutputAtIndex(index, binaryOperation(a, b));
-        }
+      this.dispatch = computeDispatch(
+          this.dispatchLayout, this.outputShape, this.workGroupSize,
+          [this.workPerThread, 1, 1]);
+      this.shaderKey =
+          `binaryShared_${op}_${lastDimensionSize}_${useSharedMemoryWithB}`;
+
+      const sharedIndexSnippet = lastDimensionSize > 1 ?
```

Comment:
The `getUserCode()`'s purpose is to hide the complexity of generating the shader body and only be called once if the `shaderKey` doesn't change. However, in this change, you break this rule and execute code L69~L140 every time when `new BinaryOpProgram`. You probably need to move them into `getUserCode` to  keep similar with before.

Reply:
I see. Please take another look at the fix.